src/models/user.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 Damir Jelić
+// Copyright 2020 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+use super::UserId;
+use crate::api::r0 as api;
+use crate::events::collections::all::{Event, RoomEvent, StateEvent};
+use crate::events::presence::{PresenceEvent, PresenceEventContent, PresenceState};
+use crate::events::room::{
+    aliases::AliasesEvent,
+    canonical_alias::CanonicalAliasEvent,
+    member::{MemberEvent, MembershipState},
+    name::NameEvent,
+};
+use crate::events::EventResult;
+use crate::identifiers::RoomAliasId;
+use crate::session::Session;
+
+use js_int::UInt;
+#[cfg(feature = "encryption")]
+use tokio::sync::Mutex;
+
+#[cfg(feature = "encryption")]
+use crate::crypto::{OlmMachine, OneTimeKeys};
+#[cfg(feature = "encryption")]
+use ruma_client_api::r0::keys::{upload_keys::Response as KeysUploadResponse, DeviceKeys};
+
+#[derive(Debug)]
+/// A Matrix room member.
+pub struct User {
+    /// The human readable name of the user.
+    pub display_name: Option<String>,
+    /// The matrix url of the users avatar.
+    pub avatar_url: Option<String>,
+    /// The presence of the user, if found.
+    pub presence: Option<PresenceState>,
+    /// The presence status message, if found.
+    pub status_msg: Option<String>,
+    /// The time, in ms, since the user interacted with the server.
+    pub last_active_ago: Option<UInt>,
+    /// If the user should be considered active.
+    pub currently_active: Option<bool>,
+    /// The events that created the state of the current user.
+    // TODO do we want to hold the whole state or just update our structures.
```

Comment:
I don't know what the right answer is. Redacting membership events is a valid thing to do in Matrix, e.g. if someone changes the display name to something offensive. If we hold the full event chain here we can easily go back to the previous state, though we are not guaranteed to have the full chain anyways. So perhaps holding on to the previous display name and avatar URL and event ids so we know when redactions mess things up is enough. Lets leave it as is for now and optimize later.